src/core/Akka/Event/EventBusUnsubscriber.cs

```diff
@@ -0,0 +1,113 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="EventBusUnsubscribers.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+//     Copyright (C) 2013-2015 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Akka.Actor;
+using Akka.Actor.Internal;
+using Akka.Util.Internal;
+
+namespace Akka.Event
+{
+    /// <summary>
+    /// INTERNAL API
+    /// 
+    /// Watches all actors which subscribe on the given eventStream, and unsubscribes them from it when they are Terminated.
+    /// 
+    /// Assumptions note:
+    ///  We do not guarantee happens-before in the EventStream when 2 threads subscribe(a) / unsubscribe(a) on the same actor,
+    /// thus the messages sent to this actor may appear to be reordered - this is fine, because the worst-case is starting to
+    /// needlessly watch the actor which will not cause trouble for the stream. This is a trade-off between slowing down
+    /// subscribe calls * because of the need of linearizing the history message sequence and the possibility of sometimes
+    /// watching a few actors too much - we opt for the 2nd choice here.
+    /// </summary>
+    class EventStreamUnsubscriber : ActorBase
+    {
+        private readonly EventStream _eventStream;
+        private readonly bool _debug;
+        private readonly ActorSystem _system;
+
+        public EventStreamUnsubscriber(EventStream eventStream, ActorSystem system, bool debug)
+        {
+            _eventStream = eventStream;
+            _system = system;
+            _debug = debug;
+           
+        }
+
+        protected override bool Receive(object message)
+        {
+            return message.Match().With<Register>(register =>
+            {
+                if (_debug)
+                    _eventStream.Publish(new Debug(this.GetType().Name, GetType(),
+                        $"watching {register.Actor} in order to unsubscribe from EventStream when it terminates"));
```

Comment:
We still have VS2013 users and compatibility with C# 5 / .NET 4.5. I don't think this is valid notation.